### PR TITLE
WindowServer: Scale Mouse movements

### PR DESCRIPTION
**Pull request by ElectrodeYT**
This fixes a bug where if the current screen had scaling enabled, the mouse would be twice as fast, or it would not "pop" out of the window if using absolute positions (eg. in Qemu).

### CODE REVIEW

**Review by alimpfard — CHANGES_REQUESTED** [review]
Please wrap the commit message at 72 characters (and not at 47)

**Comment by ElectrodeYT** [comment]
Should be done now!

**Comment by alimpfard** [comment]
Nice.
Let's just wait for some CI confirmation before merge, because I can't trust my eyes with the parentheses here :P

**Comment by tomuta** [comment]
This is definitely not working correctly when you mix two displays, one 2x scaling and one regular. I changed this code on purpose, I'll have to look into how to fix this properly.

In this example, the left screen uses 2x scaling and the right screen regular 1x, as I move the mouse steadily you can see that the cursor moves twice as fast on the screen on the right, or half as fast on the 2x screen. The speed without this change is consistent on both screens and feels more natural, with this change the speed on the 2x screen feels too slow.

cursor_speed

**Comment by ElectrodeYT** [comment]
When using one screen the speed doubles when using absolute scaling and it does not correctly clip out of one window. Are you using relative or absolute mouse movement?

**Comment by ElectrodeYT** [comment]
This might also only be a problem with bochs vga, and not virtio, but I honestly do not see how that could be the case.

**Comment by tomuta** [comment]
> When using one screen the speed doubles when using absolute scaling and it does not correctly clip out of one window. Are you using relative or absolute mouse movement?

I am using relative mouse movements, sorry. Maybe it's related to that somehow? I'm booting with `vmmouse=off` (which is the only way you can use the mouse properly with multiple screens)

**Comment by ElectrodeYT** [comment]
Ill be honest - I did not test the changes with relative mouse movement because I would have thought the same issue would happen, maybe it will work if the scaling (don't know a better term) is applied only to absolute positioning?

**Comment by tomuta** [comment]
Yeah I'm not sure what exactly the problem is or how to fix it. Bottom line though `m_cursor_location` is in virtual coordinates only, regardless of any specific scaling. When the compositor draws the cursor, `m_cursor_location` is first being offset by the screen's virtual x/y coordinate, and then the position on the actual screen is taking into account the scaling. If you take the scaling into account while calculating a new `m_cursor_location` you're essentially breaking that logic, because now you may apply the scale factor twice and the coordinates would no longer be virtual coordinates. If this makes sense.

**Comment by tomuta** [comment]
I can take a look into how to fix this later today, unless you want to get to the bottom of it before then ;)

**Comment by ElectrodeYT** [comment]
I'm going to be honest - I don't understand what you are saying (I didn't study the new code that much), but the problem is definitely noticeable - boot without vmmouse=off, enable scaling, and move the mouse to the bottom and you will notice it. I don't know the best way to record it to post here, but it isn't hard to notice either.

**Comment by ElectrodeYT** [comment]
One thing that could fix it is only applying it to absolute coordinates - if m_cursor_location is virtual only, then we are setting it to physical coords anyway

**Comment by tomuta** [comment]
If the cursor location isn't virtual coordinates then things get very tricky very quickly, e.g. when moving between screens with different scaling. I initially went down that path. By being it virtual coordinates it's much easier to determine which screen the mouse is on, and then it's just a matter of calculating the actual location on the bitmap by applying that screen's scale factor.

**Comment by tomuta** [comment]
We probably shouldn't at all care about `current_screen.physical_width()` or `current_screen.physical_height()`, as that implies scale and isn't virtual coordinates.

**Comment by ElectrodeYT** [comment]
See if this fixes it - from my understanding of what you said (that m_cursor_location is virtual only) the non-relative movemnt stuff _does_ set it to absolute coords, breaking it.
It does fix it for me, but I dont know if it would break it for multi monitor.

**Comment by ElectrodeYT** [comment]
`m_cursor_location = { packet.x * current_screen.physical_width() / 0xffff, packet.y * current_screen.physical_height() / 0xffff };`
This sets the m_cursor_location to the absolute, physical coords. You're right, we probably should't really use current_screen.physical_xxx() for this

**Comment by tomuta** [comment]
> See if this fixes it - from my understanding of what you said (that m_cursor_location is virtual only) the non-relative movemnt stuff _does_ set it to absolute coords, breaking it.
> It does fix it for me, but I dont know if it would break it for multi monitor.

This does indeed make the speed consistent. But the logic seems over-complicated because it takes into account the real resolution and then turns it back into virtual. Maybe something like this:

```c++
m_cursor_location = { (packet.x * current_screen.width() / 0xffff), (packet.y * current_screen.height() / 0xffff) };
```

But I haven't tried it, but should behave the same I think.

**Comment by ElectrodeYT** [comment]
I just tried it lol yeah it does

**Comment by ElectrodeYT** [comment]
There you go, using .width() and .height() now :)

**Comment by tomuta** [comment]
Ok, I just verified that what I suggested also works equally, I think we should go with that if you can confirm it works correctly in absolute mode. I have not tested that.

**Comment by ElectrodeYT** [comment]
It does

**Comment by tomuta** [comment]
So in hind sight, I guess I neglected to revisit the `else` branch of the `if (packet.is_relative)` ;-) Thanks for resolving this :)

**Comment by ElectrodeYT** [comment]
No problem :)

**Comment by ElectrodeYT** [comment]
CI error was a time-out, not related.

**Pull request merged**
